File: Excel/1000-1099/1010/Challenge 1010.R
library(tidyverse)
library(readxl)

path <- "1000-1099/1010/1010 Record Alignment.xlsx"
input <- read_excel(path, range = "A2:A24")
test <- read_excel(path, range = "C2:E7")

result <- input %>%
  mutate(count = cumsum(is.na(input[[1]])) + 1) %>%
  na.omit() %>%
  mutate(nr = row_number(), .by = count) %>%
  mutate(
    adj_nr = case_when(nr == 3 & !str_detect(Data, "\\d") ~ 4, TRUE ~ nr)
  ) %>%
  select(-c(nr)) %>%
  pivot_wider(names_from = adj_nr, values_from = Data) %>%
  select(-count) %>%
  unite("Full Name", c(`1`, `2`), sep = " ", na.rm = TRUE) %>%
  mutate(`3` = as.numeric(`3`)) %>%
  select(`Full Name`, Age = `3`, City = `4`)

all.equal(result, test)
# True

File: Excel/1000-1099/1010/Challenge 1010.py
import pandas as pd
import numpy as np

path = "1000-1099/1010/1010 Record Alignment.xlsx"
input = pd.read_excel(path, usecols="A", nrows=23, skiprows=1)
test = pd.read_excel(path, usecols="C:E", nrows=5, skiprows=1)

df = input.copy()
first_col = df.columns[0]
if first_col != "Data":
    df = df.rename(columns={first_col: "Data"})
df["count"] = df["Data"].isna().cumsum() + 1
df = df.dropna(subset=["Data"]).copy()
df["nr"] = df.groupby("count").cumcount() + 1
df["adj_nr"] = np.where(
    (df["nr"] == 3) & (~df["Data"].astype(str).str.contains(r"\d", na=False)),
    4,
    df["nr"],
)
wide = df.pivot(index="count", columns="adj_nr", values="Data").reset_index(drop=True)
wide["Full Name"] = (
    wide.reindex(columns=[1, 2])
    .fillna("")
    .agg(" ".join, axis=1)
    .str.replace(r"\s+", " ", regex=True)
    .str.strip()
)
wide["Age"] = pd.to_numeric(wide.get(3), errors="coerce")
wide["City"] = wide.get(4)
result = wide[["Full Name", "Age", "City"]]

print(result.equals(test))
# True
